zerver/lib/push_notifications.py

```diff
@@ -644,10 +643,12 @@ def failure_processor(event: Dict[str, Any]) -> None:
         if android_devices:
             send_android_push_notification(android_devices, gcm_payload, gcm_options)
 
-    for user_message in user_messages:
-        # TODO make this O(1) queries... including access_message, above
-        user_message.flags.active_mobile_push_notification = False
-        user_message.save(update_fields=["flags"])
+    UserMessage.objects.filter(
+        user_profile_id=user_profile_id,
+        message_id__in=message_ids,
+    ).update(
+        flags=F('flags').bitand(
+            ~UserMessage.flags.active_mobile_push_notification))
```

Comment:
This change is great.

But just an observation about how Django works: Doing a bulk update query like this will not run any post-save hooks for the table as `.save()` would.  We don't cache UserMessage rows and thus don't use any post-save hooks, but just wanted to flag this subtle semantics as a good thing to know about -- if we were doing this sort of thing with `UserProfile`, we'd need explicit cache-flushing code here :).